src/Content/ContentStorage.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+namespace Kirby\Content;
+
+use Kirby\Cms\Language;
+use Kirby\Cms\ModelWithContent;
+
+/**
+ * Interface for content storage handlers;
+ * note that it is so far not viable to build custom
+ * handlers because the CMS core relies on the filesystem
+ * and cannot fully benefit from this abstraction yet
+ * @internal
+ * @since 4.0.0
+ *
+ * @package   Kirby Content
+ * @author    Lukas Bestle <lukas@getkirby.com>
+ * @link      https://getkirby.com
+ * @copyright Bastian Allgeier
+ * @license   https://getkirby.com/license
+ */
+interface ContentStorage
+{
+	public function __construct(ModelWithContent $model);
+
```

Comment:
Not sure. Thinking about it again, I feel like `language()` as a helper method doesn't belong in this class at all, especially because the implementation won't differ between different storage handlers. Maybe we can find a better place for it.

Reply:
Ok, maybe phrased differently: can there be an implementation of `ContentStorage` that doesn't need a language method?

Reply:
No, since passing the language as object is already a requirement in the `ContentStorage` interface.

The language method is only there because the core passes around language codes as strings at the moment. With our planned "always multilang" refactoring I think that we will be able to update other places in the core to use the same logic, so the language method is not specific to the content storage at all, just used here for the first time.

Reply:
I think that’s actually an issue. We’re diverting here quite a lot from the usual Kirby way which is very focused on ease of use of its interfaces/methods/API. Which is why we allow passing a language as language code string almost everywhere. It’s not as great for type hinting as a language object, but it raises the UX (or developer X) a Lot.

With requiring a language object, the code of the language class gets leaner and easier. But it’s actually just externalizing the costs and blows up the code all around the system.

I think we should keep this quality of ease of use and allow passing a language code string and let the class handle getting the object internally. This would be more consistent with Kirby.

(It’s probably also the biggest edge I have with VersionIdentifier etc - using them internally can make a lot of sense, but externalizing the work to pass them as object instead of a simple string doesn’t feel right and match how Kirby usually deals with this and offers an easy-to-use interface, e.g. page status as string)

Reply:
On passing language codes as strings:

The language method took over its logic 1:1 from the old `$model->contentFile()` method, including the `$force` behavior that is needed for the conversion code in the `Language` class.

Where do we put all of that logic when we switch to strings for the language code arguments? I see two alternatives:

- We handle all of the complexity inside the content storage handler. This has two gotchas: a) each handler would need to implement the same logic and b) every handler method would need an additional `$force` argument, which I think would be a huge downgrade in code quality. So IMO this is not really an option.
- We handle the complexity in the calling code and only pass the "final" language code to the storage handler that can be used 1:1 for the content filename. To avoid code duplication, I'd still put the logic for the language code conversion in a helper method and we'd need to find a place for that.

Reply:
I would handle it inside. We could ourselves also handle it outside reasonably but then every developer calling these methods would need to have the same awareness etc. - they’re much more used to pass the language code string as everywhere else in the system. So I think this would offer a much better developer experience if we handle it inside. 

I’ll look closer at it tonight if there aren’t good ways to handle this inside once and not on every method. Or maybe it isn’t the biggest deal to do it on the methods how we’ve done so far too. I’ll need to do a closer take on that.

Reply:
I'd much prefer making the storage handlers as dumb as possible regarding this. They shouldn't have to think about language edge cases, just receive either a language code or `null` in single-lang.

Reply:
I know, but just externalising these complexities into the code that developers have to write can't be a good answer either. We need to find some middle ground.

Reply:
Looking at the `$force` parameter, that seems to really be the culprit. I am not Sure if it actually should exist. Or if we are just using it to not face the actual correct behavior in other parts.

So if we look at where this is used:
- `Cms\Language`: There's code in those methods that actually doesn't belong in the language class. The language class should not be concerned with moving the files. In the Language class there should probably just be one call per object, something like `$model->storage()->convertLanguage(string $from, string $to)`. And then each storage handler needs to know what to do. In our case PlainStorage to do the `F::move` calls. I could imagine that in this context the need for the $force parameter disappears.
- `Cms\ContentTranslation:: contentFile`: seems like an odd case to force this, there shouldn't be a content translation for any invalid language, that'd be a problem itself
- `Cms\Translations`: I think those two methods your PR removed already anyways

Reply:
As Nico and I discussed separately, this is something we want to improve, but outside of this PR. Nico will work on a solution. I'll leave this discussion open for reference.